### ISSUE TITLE
Warn when elided lifetime is shorter than it could have been.

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#42287]
Hi! Several times I've faced this interesting case of lifetime miselision:

```Rust
struct Foo<'a> { bar: &'a Bar}

impl<'a> Foo<'a> {
  fn get_bar(&self) -> & /* 'a */ Bar { self.bar }
}
```

http://play.integer32.com/?gist=3aba0526aff729bb9d62483a10732826&version=undefined

Here, we want to use `'a` lifetime for the returned `Bar` reference, but lifetime elision gives use lifetime of the self reference. The problem here is that code with elided lifetimes compiles flawlessly by itself, but may fail at the call site, and this is tremendously confusing, because you get a misleading compiler error in the completely wrong place! I've hit this issues several times in practice ([the latest one](https://github.com/matklad/fall/commit/73eeeb9d5f464fbdb2bb7b8f24975f4d8f586bb9#diff-036076241b912ab646c1ef74b94afccaL31)).

I wonder if we can warn about this? That is, we issue a warning if a lifetime in a returned position is elided, but the lifetime of the returned value from the function can actually be larger (alternatively, if the actual lifetime is among the input lifetimes, and is not the same as the one inferred by elision). The fix for warning would be to specify lifetimes explicitly (I suppose that "I did mean to use a shorter lifetime here" is a rather rare use-case).

**Comment by Mark-Simulacrum** [comment]
I don't think this is going to be a good fit for Rust, since I think there could be false positives. I'm going to close this issue -- feel free to file either an RFC or ask Clippy if they'd like something like this.

**Comment by leodasvacas** [comment]
@Mark-Simulacrum Any case in which this is a false positive is very special and should be documented with explicit lifetimes anyhow, this seems like a very good lint for rustc.
